### PR TITLE
feature: Logic signatures

### PR DESCRIPTION
**Pull request by tristanmenzel**

### CODE REVIEW

File: src/puya/awst_build/subroutine.py [hunk 1]
```diff
@@ -106,19 +107,34 @@ class ContractMethodInfo:
...
         context: ASTConversionModuleContext,
         func_def: mypy.nodes.FuncDef,
-        contract_method_info: ContractMethodInfo | None,
+        context_method_info: ContractMethodInfo | LogicSigMethodInfo | None,
         source_location: SourceLocation,
     ):
         super().__init__(context=context)
         func_loc = self._location(func_def)
-        self.contract_method_info = contract_method_info
+        match context_method_info:
```
Comment (tristanmenzel) [inline src/puya/awst_build/subroutine.py:127]: @achidlow is there a nicer way to do this?

File: src/puya/ir/context.py [hunk 2]
```diff
@@ -26,13 +26,15 @@ class IRBuildContext(CompileContext):
...
     contract: awst_nodes.ContractFragment | None = None
+    logic_sig: awst_nodes.LogicSignature | None = None
 
     def for_contract(self, contract: awst_nodes.ContractFragment) -> "IRBuildContextWithFallback":
         return attrs_extend(
             IRBuildContextWithFallback,
             self,
             default_fallback=contract.source_location,
             contract=contract,
+            logic_sig=None,
```
Comment (tristanmenzel) [inline src/puya/ir/context.py:37]: Probably not needed??

File: src/puya/ir/main.py [hunk 3]
```diff
@@ -172,6 +192,46 @@ def _build_ir(ctx: IRBuildContextWithFallback, contract: awst_nodes.ContractFrag
...
+    # visit call graph starting at entry point(s) to collect all references for each
+    callees = CalleesLookup(set)
+    approval_subs_srefs = SubroutineCollector.collect(ctx, start=logic_sig, callees=callees)
+
+    # construct unique Subroutine objects for each function
+    # that was referenced through either entry point
+    for func in approval_subs_srefs:
+        if func not in ctx.subroutines:
+            allow_implicits = _should_include_implicit_returns(
+                func, callees=callees[func], approval_program=logic_sig
```
Comment (tristanmenzel) [inline src/puya/ir/main.py:207]: This is wrong, should just always allow implicits

File: src/puya/ir/main.py [hunk 4]
```diff
@@ -172,6 +192,46 @@ def _build_ir(ctx: IRBuildContextWithFallback, contract: awst_nodes.ContractFrag
     return result
 
 
+def _build_logic_sig_ir(
+    ctx: IRBuildContextWithFallback, logic_sig: awst_nodes.LogicSignature
+) -> LogicSignature:
+    # visit call graph starting at entry point(s) to collect all references for each
+    callees = CalleesLookup(set)
+    approval_subs_srefs = SubroutineCollector.collect(ctx, start=logic_sig, callees=callees)
```
Comment (tristanmenzel) [inline src/puya/ir/main.py:200]: rename

**Review by achidlow — APPROVED** [review]

**Pull request merged**